classes/class-p4-search.php

```diff
@@ -43,45 +45,195 @@ public function __construct( $search_query, $templates = [ 'search.twig', 'archi
+				 */
+				return Timber::get_posts();
+			}
+
+			$args = [
+				's'              => $this->search_query,
+				'posts_per_page' => self::POSTS_LIMIT,          // Set a high maximum because -1 will get ALL posts and this can be very intensive in production.
+				'no_found_rows'  => true,                       // This means that the result counters of each filter might not be 100% precise.
+			];
+
+			foreach ( $this->filters as $type => $filter_type ) {
+				foreach ( $filter_type as $filter ) {
+					switch ( $type ) {
+						case 'cat':
+							if ( count( (array) $filter_type ) > 1 ) {
+								$args['category__and'][] = $filter['id'];
+							} else {
+								$args['tax_query'][] = [
+									[
+										'taxonomy' => 'category',
+										'field'    => 'term_id',
+										'terms'    => $filter['id'],
+									],
+								];
+							}
+							break;
+						case 'tag':
+							if ( count( (array) $filter_type ) > 1 ) {
+								$args['tag__and'][] = $filter['id'];
+							} else {
+								$args['tax_query'][] = [
+									[
+										'taxonomy' => 'post_tag',
+										'field'    => 'term_id',
+										'terms'    => $filter['id'],
+									],
+								];
+							}
+							break;
+						case 'ptype':
+							if ( count( (array) $filter_type ) > 1 ) {
+								$args['post__in'][] = $filter['id'];
+							} else {
+								$args['tax_query'][] = [
+									[
+										'taxonomy' => 'p4-page-type',
+										'field'    => 'term_id',
+										'terms'    => $filter['id'],
+									],
+								];
+							}
+							break;
+						case 'ctype':
+							switch ( $filter['id'] ) {
+								case 0:
+									$args['post_type']   = 'page';
+									$args['post_parent'] = get_page_by_path( 'act', 'OBJECT', 'page' )->ID;
+									break;
+								case 1:
+									$args['post_type'] = 'attachment';
+									break;
+								case 2:
+									$args['post_type'] = 'page';
+									$args['post_parent__not_in'][] = get_page_by_path( 'act', 'OBJECT', 'page' )->ID;
+									break;
+								case 3:
+									$args['post_type'] = 'post';
+									break;
+							}
+							break;
+					}
+				}
+			}
+
+			/*
+			 * 1. Pass params for SWP_Query and get posts.
+			 * 2. If more params that are not supported by SWP_Query like post_parent, post_parent__not_in,
+			 *    tag__and, category__and, are required then pass them to WP_Query and get posts.
+			 * 3. Get the respective Timber Posts, so that we can use Timber functionality in our search template.
+			 */
+			$posts = ( new SWP_Query( $args ) )->posts;
+
+			if ( 'attachment' !== $args['post_type'] ) {    // This does not happen when filtering for attachments, since WP_Query does not support searching within rich text documents.
+				$ids = [];
+				foreach ( $posts as $post ) {
+					$ids[] = $post->ID;
+				}
+				$args['post__in'] = $ids;
+				$args['orderby']  = 'post__in';
+				$posts            = ( new WP_Query( $args ) )->posts;
 			}
+
+			$timber_posts = [];
```

Comment:
Maybe it is better to move the conversion of WP_Post objects to TimberPost objects in a different function to keep the function from serving several purposes. Because search is a vital component of the app (and probably it will get more complex in the future) we should implement tests at some point. So I think is better to keep the functions as small as possible, to server a single concern.